sklearn/preprocessing/data.py

```diff
@@ -2567,3 +2571,265 @@ def quantile_transform(X, axis=0, n_quantiles=1000,
+                                     "supported")
+
+        X = check_array(X, dtype=np.object, copy=True)
+        n_samples, n_features = X.shape
+
+        self._label_encoders_ = [LabelEncoder() for _ in range(n_features)]
+
+        for i in range(n_features):
+            le = self._label_encoders_[i]
+            Xi = X[:, i]
+            if self.categories == 'auto':
+                le.fit(Xi)
+            else:
+                valid_mask = np.in1d(Xi, self.categories[i])
+                if not np.all(valid_mask):
+                    if self.handle_unknown == 'error':
+                        diff = np.unique(Xi[~valid_mask])
+                        msg = ("Found unknown categories {0} in column {1}"
+                               " during fit".format(diff, i))
+                        raise ValueError(msg)
+                le.classes_ = np.array(self.categories[i])
+
+        self.categories_ = [le.classes_ for le in self._label_encoders_]
+
+        return self
+
+    def transform(self, X):
+        """Transform X using specified encoding scheme.
+
+        Parameters
+        ----------
+        X : array-like, shape [n_samples, n_features]
+            The data to encode.
+
+        Returns
+        -------
+        X_out : sparse matrix or a 2-d array
+            Transformed input.
+
+        """
+        X = check_array(X, dtype=np.object, copy=True)
+        n_samples, n_features = X.shape
+        X_int = np.zeros_like(X, dtype=np.int)
+        X_mask = np.ones_like(X, dtype=np.bool)
+
+        for i in range(n_features):
+            valid_mask = np.in1d(X[:, i], self.categories_[i])
+
+            if not np.all(valid_mask):
+                if self.handle_unknown == 'error':
+                    diff = np.unique(X[~valid_mask, i])
+                    msg = ("Found unknown categories {0} in column {1}"
+                           " during transform".format(diff, i))
+                    raise ValueError(msg)
+                else:
+                    # Set the problematic rows to an acceptable value and
+                    # continue `The rows are marked `X_mask` and will be
+                    # removed later.
+                    X_mask[:, i] = valid_mask
+                    X[:, i][~valid_mask] = self.categories_[i][0]
+            X_int[:, i] = self._label_encoders_[i].transform(X[:, i])
+
+        if self.encoding == 'ordinal':
+            return X_int.astype(self.dtype, copy=False)
+
+        mask = X_mask.ravel()
+        n_values = [cats.shape[0] for cats in self.categories_]
+        n_values = np.array([0] + n_values)
+        indices = np.cumsum(n_values)
+
+        column_indices = (X_int + indices[:-1]).ravel()[mask]
+        row_indices = np.repeat(np.arange(n_samples, dtype=np.int32),
+                                n_features)[mask]
+        data = np.ones(n_samples * n_features)[mask]
+
+        out = sparse.csc_matrix((data, (row_indices, column_indices)),
```

Comment:
I am not fully sure why I wrote this here, but should this be a csr matrix rather than csc ?

Reply:
I need dinner before I can answer that question. brb.

Reply:
Idea for this one?

Reply:
the way you construct it with ``(data, (i, j))`` it will always construct a COO matrix first. So the current code will construct a COO matrix, then convert it to a CSC, and then you convert it to a CSR.
Either call the CSR constructor this way, or provide the default CSR construction with indptr. Is ``indptr`` just ``mask.sum(axis=1).cumsum()``?

Reply:
Sorry, ``X_mask.sum(axis=1).cumsum()`` that's probably the better way to construct this. (you need to add a leading zero, though)

Reply:
Ah, I missed I actually already converted to crs after construction. So I updated the construction code a bit to directly construct a CSR matrix